native/dsteqr.go

```diff
@@ -0,0 +1,394 @@
+		icompz = 1
+	} else if compz == lapack.EigBoth {
+		icompz = 2
+	}
+
+	if n == 0 {
+		return true
+	}
+	if n == 1 {
+		if icompz == 2 {
+			z[0] = 1
+		}
+		return true
+	}
+
+	bi := blas64.Implementation()
+
+	eps := dlamchE
+	eps2 := eps * eps
+	safmin := dlamchS
+	safmax := 1 / safmin
+	ssfmax := math.Sqrt(safmax) / 3
+	ssfmin := math.Sqrt(safmin) / eps2
+
+	// Compute the eigenvalues and eigenvectors of the tridiagonal matrix.
+	if icompz == 2 {
+		impl.Dlaset(blas.All, n, n, 0, 1, z, ldz)
+	}
+	const maxit = 30
+	nmaxit := n * maxit
+
+	jtot := 0
+
+	// Determine where the matrix splits and choose QL or QR iteration for each
+	// block, according to whether top or bottom diagonal element is smaller.
+	// TODO(btracey): Reformat the gotos to use structural flow techniques.
+	// TODO(btracey): Move these variables closer to actual usage when
+	// gotos are removed.
+	l1 := 0
+	nm1 := n - 1
+	var l, lend, lendsv, lm1, lsv, m, mm1 int
+	var anorm, c, g, p, r, rt1, rt2, s float64
+
+	type scaletype int
+	const (
+		none scaletype = iota
+		down
+		up
+	)
+	var iscale scaletype
+
+Ten:
+	if l1 > n-1 {
+		goto OneSixty
+	}
+	if l1 > 0 {
+		e[l1-1] = 0
+	}
+	if l1 <= nm1 {
+		for m = l1; m < nm1; m++ {
+			test := math.Abs(e[m])
+			if test == 0 {
+				goto Thirty
+			}
+			if test <= (math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1])))*eps {
+				e[m] = 0
+				goto Thirty
+			}
+		}
+	}
+	m = n - 1
+Thirty:
+	l = l1
+	lsv = l
+	lend = m
+	lendsv = lend
+	l1 = m + 1
+	if lend == 0 {
+		goto Ten
+	}
+
+	// Scale submatrix in rows and columns L to Lend
+	anorm = impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+	switch {
+	case anorm == 0:
+		goto Ten
+	case anorm > ssfmax:
+		iscale = down
+		// TODO(btracey): Why is lda n?
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l, 1, e[l:], n)
+	case anorm < ssfmin:
+		iscale = up
+		// TODO(btracey): Why is lda n?
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l, 1, e[l:], n)
+	}
+
+	// Choose between QL and QR.
+	if math.Abs(d[lend]) < math.Abs(d[l]) {
+		lend = lsv
+		l = lendsv
+	}
+	if lend > l {
+		// QL Iteration. Look for small subdiagonal element.
+	Fourty:
+		if l != lend {
+			lendm1 := lend - 1
+			for m = l; m <= lendm1; m++ {
+				v := math.Abs(e[m])
+				if v*v <= (eps2*math.Abs(d[m]))*math.Abs(d[m+1])+safmin {
+					goto Sixty
+				}
+			}
+		}
+		m = lend
+	Sixty:
+		if m < lend {
+			e[m] = 0
+		}
+		p = d[l]
+		if m == l {
+			goto Eighty
+		}
+
+		// If remaining matrix is 2×2, use Dlae2 to compute its eigensystem.
+		if m == l+1 {
+			if icompz > 0 {
+				rt1, rt2, work[l], work[n-1+l] = impl.Dlaev2(d[l], e[l], d[l+1])
+				impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward,
+					n, 2, work[l:], work[n-1+l:], z[l:], ldz)
+			} else {
+				rt1, rt2 = impl.Dlae2(d[l], e[l], d[l+1])
+			}
+			d[l] = rt1
```

Comment:
It seems to me that `rt1` and `rt2` are not necessary, `d[l]` and `d[l+1]` can be assigned to directly.